File: app/src/main/java/com/xyoye/dandanplay/ui/weight/expandable/ExpandableLayout.java
package com.xyoye.dandanplay.ui.weight.expandable;

import android.animation.Animator;
import android.animation.ValueAnimator;
import android.content.Context;
import android.content.res.Configuration;
import android.content.res.TypedArray;
import android.os.Bundle;
import android.os.Parcelable;
import android.util.AttributeSet;
import android.view.View;
import android.view.animation.Interpolator;
import android.view.animation.LinearInterpolator;
import android.widget.LinearLayout;

import com.xyoye.dandanplay.R;

import skin.support.widget.SkinCompatFrameLayout;

/**
 * copyright https://github.com/cachapa/ExpandableLayout
 */

public class ExpandableLayout extends SkinCompatFrameLayout {
    private static ExpandableLayout cacheLayout;

    public interface State {
        int COLLAPSED = 0;
        int COLLAPSING = 1;
        int EXPANDING = 2;
        int EXPANDED = 3;
    }

    public static final String KEY_SUPER_STATE = "super_state";
    public static final String KEY_EXPANSION = "expansion";

    public static final int HORIZONTAL = 0;
    public static final int VERTICAL = 1;

    private static final int DEFAULT_DURATION = 300;

    private int duration = DEFAULT_DURATION;
    private float parallax;
    private float expansion;
    private int orientation;
    private int state;

    private Interpolator interpolator = new LinearInterpolator();
    private ValueAnimator animator;

    private OnExpansionUpdateListener listener;

    public ExpandableLayout(Context context) {
        this(context, null);
    }

    public ExpandableLayout(Context context, AttributeSet attrs) {
        super(context, attrs);

        if (attrs != null) {
            TypedArray a = getContext().obtainStyledAttributes(attrs, R.styleable.ExpandableLayout);
            duration = a.getInt(R.styleable.ExpandableLayout_el_duration, DEFAULT_DURATION);
            expansion = a.getBoolean(R.styleable.ExpandableLayout_el_expanded, false) ? 1 : 0;
            orientation = a.getInt(R.styleable.ExpandableLayout_android_orientation, VERTICAL);
            parallax = a.getFloat(R.styleable.ExpandableLayout_el_parallax, 1);
            a.recycle();

            state = expansion == 0 ? State.COLLAPSED : State.EXPANDED;
            setParallax(parallax);
        }
    }

    @Override
    protected Parcelable onSaveInstanceState() {
        Parcelable superState = super.onSaveInstanceState();
        Bundle bundle = new Bundle();

        expansion = isExpanded() ? 1 : 0;

        bundle.putFloat(KEY_EXPANSION, expansion);
        bundle.putParcelable(KEY_SUPER_STATE, superState);

        return bundle;
    }

    @Override
    protected void onRestoreInstanceState(Parcelable parcelable) {
        Bundle bundle = (Bundle) parcelable;
        expansion = bundle.getFloat(KEY_EXPANSION);
        state = expansion == 1 ? State.EXPANDED : State.COLLAPSED;
        Parcelable superState = bundle.getParcelable(KEY_SUPER_STATE);

        super.onRestoreInstanceState(superState);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);

        int width = getMeasuredWidth();
        int height = getMeasuredHeight();

        int size = orientation == LinearLayout.HORIZONTAL ? width : height;

        setVisibility(expansion == 0 && size == 0 ? GONE : VISIBLE);

        int expansionDelta = size - Math.round(size * expansion);
        if (parallax > 0) {
            float parallaxDelta = expansionDelta * parallax;
            for (int i = 0; i < getChildCount(); i++) {
                View child = getChildAt(i);
                if (orientation == HORIZONTAL) {
                    int direction = -1;
                    if (getLayoutDirection() == LAYOUT_DIRECTION_RTL) {
                        direction = 1;
                    }
                    child.setTranslationX(direction * parallaxDelta);
                } else {
                    child.setTranslationY(-parallaxDelta);
                }
            }
        }

        if (orientation == HORIZONTAL) {
            setMeasuredDimension(width - expansionDelta, height);
        } else {
            setMeasuredDimension(width, height - expansionDelta);
        }
    }

    @Override
    protected void onConfigurationChanged(Configuration newConfig) {
        if (animator != null) {
            animator.cancel();
        }
        super.onConfigurationChanged(newConfig);
    }

    /**
     * Get expansion state
     *
     * @return one of {@link State}
     */
    public int getState() {
        return state;
    }

    public boolean isExpanded() {
        return state == State.EXPANDING || state == State.EXPANDED;
    }

    public void toggle() {
        toggle(true);
    }

    public void toggle(boolean animate) {
        if (isExpanded()) {
            if (cacheLayout != this) {
                collapse(animate);
            }
            if (cacheLayout != null) {
                cacheLayout.collapse();
                cacheLayout = null;
            }
        } else {
            if (cacheLayout != null) {
                cacheLayout.collapse();
            }
            expand(animate);
            cacheLayout = this;
        }
    }

    public void expand() {
        expand(true);
    }

    public void expand(boolean animate) {
        setExpanded(true, animate);
    }

    public void collapse() {
        collapse(true);
    }

    public void collapse(boolean animate) {
        setExpanded(false, animate);
    }

    /**
     * Convenience method - same as calling setExpanded(expanded, true)
     */
    public void setExpanded(boolean expand) {
        setExpanded(expand, true);
    }

    public void setExpanded(boolean expand, boolean animate) {
        if (expand == isExpanded()) {
            return;
        }

        int targetExpansion = expand ? 1 : 0;
        if (animate) {
            animateSize(targetExpansion);
        } else {
            setExpansion(targetExpansion);
        }
    }

    public int getDuration() {
        return duration;
    }

    public void setInterpolator(Interpolator interpolator) {
        this.interpolator = interpolator;
    }

    public void setDuration(int duration) {
        this.duration = duration;
    }

    public float getExpansion() {
        return expansion;
    }

    public void setExpansion(float expansion) {
        if (this.expansion == expansion) {
            return;
        }

        // Infer state from previous value
        float delta = expansion - this.expansion;
        if (expansion == 0) {
            state = State.COLLAPSED;
        } else if (expansion == 1) {
            state = State.EXPANDED;
        } else if (delta < 0) {
            state = State.COLLAPSING;
        } else if (delta > 0) {
            state = State.EXPANDING;
        }

        setVisibility(state == State.COLLAPSED ? GONE : VISIBLE);
        this.expansion = expansion;
        requestLayout();

        if (listener != null) {
            listener.onExpansionUpdate(expansion, state);
        }
    }

    public float getParallax() {
        return parallax;
    }

    public void setParallax(float parallax) {
        // Make sure parallax is between 0 and 1
        parallax = Math.min(1, Math.max(0, parallax));
        this.parallax = parallax;
    }

    public int getOrientation() {
        return orientation;
    }

    public void setOrientation(int orientation) {
        if (orientation < 0 || orientation > 1) {
            throw new IllegalArgumentException("Orientation must be either 0 (horizontal) or 1 (vertical)");
        }
        this.orientation = orientation;
    }

    public void setOnExpansionUpdateListener(OnExpansionUpdateListener listener) {
        this.listener = listener;
    }

    private void animateSize(int targetExpansion) {
        if (animator != null) {
            animator.cancel();
            animator = null;
        }

        animator = ValueAnimator.ofFloat(expansion, targetExpansion);
        animator.setInterpolator(interpolator);
        animator.setDuration(duration);

        animator.addUpdateListener(new ValueAnimator.AnimatorUpdateListener() {
            @Override
            public void onAnimationUpdate(ValueAnimator valueAnimator) {
                setExpansion((float) valueAnimator.getAnimatedValue());
            }
        });

        animator.addListener(new ExpansionListener(targetExpansion));

        animator.start();
    }

    @Override
    protected void onDetachedFromWindow() {
        if (cacheLayout != null) {
            cacheLayout = null;
        }
        super.onDetachedFromWindow();

    }

    public interface OnExpansionUpdateListener {
        /**
         * Callback for expansion updates
         *
         * @param expansionFraction Value between 0 (collapsed) and 1 (expanded) representing the the expansion progress
         * @param state             One of {@link State} repesenting the current expansion state
         */
        void onExpansionUpdate(float expansionFraction, int state);
    }

    private class ExpansionListener implements Animator.AnimatorListener {
        private int targetExpansion;
        private boolean canceled;

        public ExpansionListener(int targetExpansion) {
            this.targetExpansion = targetExpansion;
        }

        @Override
        public void onAnimationStart(Animator animation) {
            state = targetExpansion == 0 ? State.COLLAPSING : State.EXPANDING;
        }

        @Override
        public void onAnimationEnd(Animator animation) {
            if (!canceled) {
                state = targetExpansion == 0 ? State.COLLAPSED : State.EXPANDED;
                setExpansion(targetExpansion);
            }
        }

        @Override
        public void onAnimationCancel(Animator animation) {
            canceled = true;
        }

        @Override
        public void onAnimationRepeat(Animator animation) {
        }
    }
}

File: user_component/src/main/java/com/xyoye/user_component/ui/weight/ExpandableLayout.kt
package com.xyoye.user_component.ui.weight

import android.animation.Animator
import android.animation.ValueAnimator
import android.content.Context
import android.content.res.Configuration
import android.os.Bundle
import android.os.Parcelable
import android.util.AttributeSet
import android.view.animation.LinearInterpolator
import android.widget.FrameLayout
import android.widget.LinearLayout
import androidx.core.view.isGone
import androidx.core.view.isVisible
import com.xyoye.user_component.R
import kotlin.math.max
import kotlin.math.min
import kotlin.math.round

/**
 * Created by xyoye on 2021/3/4.
 *
 * copyright https://github.com/cachapa/ExpandableLayout
 */

class ExpandableLayout : FrameLayout {

    enum class State {
        COLLAPSED,
        COLLAPSING,
        EXPANDING,
        EXPANDED
    }

    companion object {
        @JvmStatic
        private var cacheLayout: ExpandableLayout? = null

        private const val DEFAULT_DURATION = 300

        private const val HORIZONTAL = 0
        private const val VERTICAL = 1

        private const val KEY_EXPANSION = "key_expansion"
        private const val KEY_SUPER_STATE = "key_super_state"
    }

    var mDuration = 0
    var mExpansion = 0f
    private var orientation = VERTICAL
    private var parallax = 0f

    private var currentState = State.COLLAPSED
    var mAnimator: ValueAnimator? = null
    var mInterpolator = LinearInterpolator()

    private var mExpansionUpdateBlock: ((Float, State) -> Unit)? = null

    constructor(context: Context) : this(context, null)
    constructor(context: Context, attrs: AttributeSet?) : this(context, attrs, 0)
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    ) {
        attrs?.apply {
            val typedArray = context.obtainStyledAttributes(attrs, R.styleable.ExpandableLayout)
            mDuration = typedArray.getInt(R.styleable.ExpandableLayout_duration, DEFAULT_DURATION)
            val expanded = typedArray.getBoolean(R.styleable.ExpandableLayout_expanded, false)
            orientation =
                typedArray.getInt(R.styleable.ExpandableLayout_android_orientation, VERTICAL)
            parallax = typedArray.getFloat(R.styleable.ExpandableLayout_parallax, 0f)
            typedArray.recycle()

            mExpansion = if (expanded) 1f else 0f
            currentState = if (expanded) State.EXPANDED else State.COLLAPSED
            setParallax(parallax)
        }
    }

    override fun onSaveInstanceState(): Parcelable {
        val superState = super.onSaveInstanceState()
        return Bundle().apply {
            mExpansion = if (isExpanded()) 1f else 0f
            putFloat(KEY_EXPANSION, mExpansion)
            putParcelable(KEY_SUPER_STATE, superState)
        }
    }

    override fun onRestoreInstanceState(state: Parcelable?) {
        if (state != null && state is Bundle) {
            mExpansion = state.getFloat(KEY_EXPANSION)
            currentState = if (mExpansion == 1f) State.EXPANDED else State.COLLAPSED
            val superSate: Parcelable? = state.getParcelable(KEY_SUPER_STATE)
            super.onRestoreInstanceState(superSate)
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)

        val size = if (orientation == LinearLayout.HORIZONTAL) measuredWidth else measuredHeight
        isGone = mExpansion == 0f && size == 0

        val expansionDelta = size - round(size * mExpansion)
        if (parallax > 0) {
            val parallaxDelta = expansionDelta * parallax
            for (index in 0..childCount) {
                val childView = getChildAt(index)
                if (orientation == HORIZONTAL) {
                    val direction = if (layoutDirection == LAYOUT_DIRECTION_RTL) 1 else -1
                    childView.translationX = direction * parallaxDelta
                } else {
                    childView.translationY = -parallaxDelta
                }

            }
        }

        if (orientation == HORIZONTAL) {
            setMeasuredDimension((measuredWidth - expansionDelta).toInt(), measuredHeight)
        } else {
            setMeasuredDimension(measuredWidth, (measuredHeight - expansionDelta).toInt())
        }
    }

    override fun onConfigurationChanged(newConfig: Configuration?) {
        mAnimator?.cancel()
        super.onConfigurationChanged(newConfig)
    }

    fun getCurrentState() = currentState

    fun isExpanded(): Boolean {
        return currentState == State.EXPANDING || currentState == State.EXPANDED
    }

    fun toggle(animate: Boolean = true) {
        cacheLayout = if (isExpanded()) {
            if (cacheLayout != this) {
                collapse(animate)
            }
            cacheLayout?.collapse()
            null
        } else {
            cacheLayout?.collapse()
            expand(animate)
            this
        }
    }

    fun collapse(animate: Boolean = true) {
        setExpanded(false, animate)
    }

    fun setExpansionObserver(block: (expansionFraction: Float, state: State) -> Unit) {
        mExpansionUpdateBlock = block
    }

    fun expand(animate: Boolean = true) {
        setExpanded(true, animate)
    }

    fun setParallax(parallaxValue: Float) {
        var value = max(0f, parallaxValue)
        value = min(1f, value)
        parallax = value
    }

    fun getParallax() = parallax

    private fun setExpanded(expanded: Boolean, animate: Boolean) {
        if (expanded == isExpanded()) {
            return
        }
        val targetExpansion = if (expanded) 1f else 0f
        if (animate) {
            animeSize(targetExpansion)
        } else {
            setExpansion(targetExpansion)
        }
    }

    private fun animeSize(targetExpansion: Float) {
        mAnimator?.cancel()

        mAnimator = ValueAnimator.ofFloat(mExpansion, targetExpansion).apply {
            interpolator = mInterpolator
            duration = mDuration.toLong()
            addUpdateListener {
                setExpansion(it.animatedValue as Float)
            }
            addListener(ExpansionListener(targetExpansion))
            start()
        }
    }

    private fun setExpansion(expansion: Float) {
        if (expansion == mExpansion) {
            return
        }

        val delta = expansion - mExpansion
        currentState = when {
            delta == 0f -> State.COLLAPSED
            delta == 1f -> State.EXPANDED
            delta < 0 -> State.COLLAPSING
            delta > 0 -> State.EXPANDING
            else -> State.COLLAPSED
        }

        isGone = currentState == State.COLLAPSED
        mExpansion = expansion
        requestLayout()
        mExpansionUpdateBlock?.invoke(expansion, currentState)
    }

    private inner class ExpansionListener(val targetExpansion: Float) : Animator.AnimatorListener {
        private var canceled = false

        override fun onAnimationStart(animation: Animator?) {
            currentState = if (targetExpansion == 0f) State.COLLAPSED else State.EXPANDING
        }

        override fun onAnimationEnd(animation: Animator?) {
            if (!canceled) {
                currentState = if (targetExpansion == 0f) State.COLLAPSED else State.EXPANDED
                setExpansion(targetExpansion)
            }
        }

        override fun onAnimationCancel(animation: Animator?) {
            canceled = true
        }

        override fun onAnimationRepeat(animation: Animator?) {
        }

    }

}